### ISSUE TITLE
Disable UDP Packet Payload Parsing

### ISSUE THREAD

**Issue by krs43** [chmorgan/packetnet#103]
We've run in to cases where we are getting sent proprietary UDP packets on the same port used by L2TP (1701). The result is that the UdpPacket intercepts the raw bytes and instead attempts to parse them as an L2tp Packet. We need a way to disable automatic Udp PayloadPacket (force payload data). Is there a way to do this already? I am working on a custom build that just disables this, but would be great if there is an API flag somewhere to optionally disable this.

**Comment by chmorgan** [comment]
Hmm.

I’d be interested in ideas for adding such an option or alternate api. How
far down the chain are you looking to parse before stopping? How can we
make it usable by others to stop at intermediate levels, that might be the
biggest open question in my mind.

Chris

**Comment by krs43** [comment]
I agree - alternatively, we could set payload data even if we set payload packet. That way if it keeps parsing down the line, when I extract UDPPacket, I can still get the payload data and parse with our own protocol

**Comment by chmorgan** [comment]
@krs43 can you try using the parent packet's HeaderDataSegment? For example:

var headerDataSegment = parentPacket.HeaderDataSegment;
var payloadByteArraySegment = header.NextSegment();

This almost calls for a new method or property in Packet like:

public ByteArraySegment PayloadRaw() // or something like this. We already have PayloadData and PayloadDataSegment but these require that the payload isn't a class instance but rather raw data.

We could also consider breaking the agreement on these properties but that's a definite api breakage. @Phyxion thoughts here?

**Comment by krs43** [comment]
I'll take a look at that - clever way of getting the next payload segment.

I think a PayloadRaw call would be great - fits with the lazy evaluation as once you have the parent frame you need, you know you can get the raw bytes and feed into custom parsers.

**Comment by chmorgan** [comment]
@krs43 let me know if it works for you, if so that's a good test case for a future property.

**Comment by Phyxion** [comment]
This is pretty much already possible. Just don't call the payload properties and it won't be parsed. If you need the payload data you can manually construct the payload segment based on the header indeed.

If you don't want to construct another segment you can also reuse Header and iterate it past its own size until the end. You could create an extension method for that for convenience. I'm using a similar structure to easily access a part of the payload.

**Comment by chmorgan** [comment]
@Phyxion thoughts on what that helper method might look like? Similar to the one we use to extract types?

**Comment by krs43** [comment]
It appears that it would work, except I pass around a handle to the UdpPacket type everywhere, so each method is recalculating the segment. I'll wrap this for now, but the PayloadRaw api would be great.

**Comment by Phyxion** [comment]
The segment is already very lightweight to construct. If you really need to then you could add a little wrapper around it to cache the segment, although I don't think you're really going to notice the difference. Furthermore, you could also use the HeaderSegment as mentioned above and add a few extension methods.
On Jul 21, 2019, at 08:04, krs43 <notifications@github.com<mailto:notifications@github.com>> wrote:

It appears that it would work, except I pass around a handle to the UdpPacket type everywhere, so each method is recalculating the segment. I'll wrap this for now, but the PayloadRaw api would be great.

**Comment by chmorgan** [comment]
@Phyxion any more details on how we can address the concern to enable the behavior @krs43 is looking for?

**Comment by Phyxion** [comment]
I stand by my previous comment on this matter 😉

**Comment by chmorgan** [comment]
You mentioned an extension method? It would be a cool feature to make it
easier for users without having to understand the details of internals.

**Comment by Phyxion** [comment]
I meant for him to create an extension method. I'm not against adding a property to get the raw payload though, but I would like to avoid an additional field for it as it would need to be allocated for each arriving packet and this is a rather limited use case. It's also already possible to use the parsed packet and then access the `BytesSegment` which is the data of the packet and any payload.

**Comment by chmorgan** [comment]
@krs43 if you were able to create and propose an extension method for raw packet handling it looks like it would make sense to integrate that into the library. Let me know if you plan to do so and need help with anything.

**Comment by krs43** [comment]
@chmorgan I never did re-wrap it as an extension method - just worked around it with a custom fork. Let me know if you end up wrapping this as an extension method, or adding a rawapi, would be great!

**Comment by Phyxion** [comment]
Keep in mind that if you fork it you need to make the source available for LGPL. Not that it's really necessary as explained above; getting the payload is trivial with HeaderSegment and not calling the Payload properties doesn't create them.

It would be nice if Packet.Net would be relicensed under MIT, this is much more flexible as almost everything is MIT now. What do you think @chmorgan?

**Comment by chmorgan** [comment]
@Phyxion part of the reason its lgpl is to allow companies to use the library but also to make it clear changes have to be contributed back. Any change contributed back via MIT would also be permitted under lgpl but the opposite might not necessarily be the case. It's worth considering of course but I'd like to hear a bit more about the benefits in the change.

**Comment by Phyxion** [comment]
That's certainly true, but it can also hinder it as many developers won't use it because of the license and its restrictions. There are so many libraries with MIT that have a ton of contributors, simply because it's easier to use the library directly rather than maintaining a fork of it. A middle ground would be the Apache 2.0 license, that gives more freedom to the users of the library but changes must still be documented and it also has some advantages to the maintainers such as patent rights. Thinking about it, this is probably a better license for Packet.Net than MIT.

**Comment by jgaulon** [comment]
In my [PR 143](https://github.com/chmorgan/packetnet/pull/143), I have introduced 2 static mapping <port/type> lists for Udp Payload decoding (to mimic the wireshark decode as.. feature). By clearing this list before parsing the packets, you will disable the L2TP payload decoding.

**Comment by PhyxionNL** [comment]
This is now possible with wring a custom decoder as per #144. Closing.
